File: 08-streaming/8.4-structured/src/main/scala/structured/WordCountWindow.scala
/*
1. Simulate a source connection
     $   ncat -l -k -p 10000

2. Run the app
    $  $SPARK_HOME/bin/spark-submit  --master local[2]   --driver-class-path logging/  \
       --class structured.WordCountWindow     target/scala-2.12/structured-streaming_2.12-1.0.jar 
 */


package structured


import org.apache.spark.sql.functions.{current_timestamp, window, lit}
import org.apache.spark.sql.SparkSession
import java.sql.Timestamp


object WordCountWindow {
  def main(args: Array[String]) {

    val spark = SparkSession.builder.appName("Structured Streaming Word Count").
                getOrCreate()
    import spark.implicits._
    println("### Spark UI available on port : " + spark.sparkContext.uiWebUrl.get.split(':')(2))

    // we can include timestamp like this also 
    //          .option('includeTimestamp', 'true')\
    val lines = spark.readStream.
                      format("socket").
                      option("host", "localhost").
                      option("port", 10000).
                      load().
                      withColumn ("timestamp", current_timestamp)
                   
    println ("Lines schema:")
    lines.printSchema
      
    // Print out incoming text
    // the 'withColumn' is for debug, so we can identify query outupt
    val query1 = lines
                .withColumn ("query", lit("query1-lines"))
                .writeStream
                .outputMode("append")
                .option("truncate", false)
                .format("console")
                .queryName("query1-lines")
                .start()
    
    // Split the lines into words, retaining timestamps
    val words = lines.as[(String, Timestamp)].flatMap(line =>
                  line._1.split(" +").map(word => (word, line._2))
                ).toDF("word", "timestamp")
    println ("words schema:")
    words.printSchema
      
    val query2 = words
                    .withColumn ("query", lit("query2-words"))
                    .writeStream
                    .outputMode("append")
                    .option("truncate", false)
                    .format("console")
                    .queryName("query2-words")
                    .start()
    
      
    // Window function is : window( column, window_interval,  slide_interval)  
    val windowedCounts = words.groupBy (
        window($"timestamp", "10 seconds"),  // window (column, window_interval)
        // window($"timestamp", "10 seconds", "5 seconds"),  // window (column, window_interval, slide_interval)
        $"word")
      .count()
      .orderBy("window")
      
    val query3 = windowedCounts
                    .withColumn ("query", lit("query3-windowedCounts"))
                    .writeStream
                    .outputMode("complete")
                    .option("truncate", false)
                    .format("console")
                    .queryName("query3-windowedCounts")
                    .start()

        // wait forever until user terminate manually
    spark.streams.awaitAnyTermination() 
    spark.stop()
  }
}

File: 08-streaming/8.4-structured/python/word_count_window.py
"""

1. Simulating a source connection
     $   ncat -l -k -p 10000

2. Run the app
     $  $SPARK_HOME/bin/spark-submit --master local[2] \
            --driver-class-path ../logging/ \
            word_count_window.py
"""

from pyspark.sql import SparkSession
from pyspark.sql.functions import lit, explode, split, current_timestamp, window


# Initialize Spark
spark = SparkSession \
    .builder \
    .appName("Structed Streaming ") \
    .getOrCreate()

# Set loglevel to Error
spark.sparkContext.setLogLevel("ERROR")

print('### Spark UI available on port : ' + spark.sparkContext.uiWebUrl.split(':')[2])


lines = spark \
    .readStream \
    .format("socket") \
    .option("host", "localhost") \
    .option("port", 10000) \
    .option('includeTimestamp', 'true')\
    .load() 

## Or we can include timestamp like this too
# lines = spark \
#     .readStream \
#     .format("socket") \
#     .option("host", "localhost") \
#     .option("port", 10000) \
#     .load() \
#     .withColumn ("timestamp", current_timestamp())

lines.printSchema()

## print out all incoming data
query1 = lines \
    .withColumn ("query", lit("query1-lines")) \
    .writeStream \
    .outputMode("append") \
    .option("truncate", False) \
    .format("console") \
    .queryName("query1-lines") \
    .start()


# Split the lines into words, retaining timestamps
# split() splits each line into an array, and explode() turns the array into multiple rows
words = lines.select(
    explode(split(lines.value, " +")).alias('word'),
    lines.timestamp
)

words.printSchema()


## print out words
query2 = words \
    .withColumn ("query", lit("query2-words")) \
    .writeStream \
    .outputMode("append") \
    .option("truncate", False) \
    .format("console") \
    .queryName("query2-words") \
    .start()

# # Window function is : window( column, window_interval,  slide_interval)  
windowdCounts = words.groupBy (
    window(words.timestamp, "10 seconds"),  # window (column, window_interval)
    # window(words.timestamp, "10 seconds", "5 seconds"),  # window (column, window_interval, slide_interval)
    words.word) \
    .count() \
    .orderBy("window") # optional



# # Start running the query that prints the running counts to the console
query3 = windowdCounts \
    .withColumn ("query", lit("query3-windowCounts")) \
    .writeStream \
    .outputMode("complete") \
    .option("truncate", False) \
    .format("console") \
    .queryName("query3-windowCounts") \
    .start()


# wait forever until user terminate manually
spark.streams.awaitAnyTermination()

spark.stop()
